### ISSUE TITLE
Missing $TMPDIR Environment variable triggers `berks vendor` bug

### ISSUE THREAD

**Issue by mreeves1** [berkshelf/berkshelf#1544]
**tl;dr** If the $TMPDIR variable is not set the `berks vendor` command creates temp directories in the working directory which will trigger an error. 

When it gets to the main application (my_app in the example below) it fails with the error:
```
"fileutils.rb:1366:in `copy': cannot copy directory /opt/jenkins-storage/workspace/MY_REPO/chef-kitchen/my_cookbooks/my_app/vendor20160418-25818-63rdy0 to itself /opt/jenkins-storage/workspace/MY_REPO/chef-kitchen/my_cookbooks/my_app/vendor20160418-25818-63rdy0/my_app/vendor20160418-25818-63rdy0 (ArgumentError)"
```

A directory starting with "vendor" is created [here](https://github.com/berkshelf/berkshelf/blob/master/lib/berkshelf/berksfile.rb#L628) and is attempting to copy into itself. 

I can think of a few solutions but not sure which people would prefer:

1. Throw an error or warning if a temporary directory is not specified
1. Save to ~/.berkshelf/tmp rather than the current working directory
1. Ignore that temporary directory in this method
1. ??? 
 
This drove me nuts because $TMPDIR exists on my Macbook but no temp directories ($TMP, $TEMP, $TEMPDIR) appear to be defined in ubuntu (12.04 at least) so I had this issue only on our build machine.

Incidentally it was ChefSpec that was calling `berks vendor`, making it even tougher to debug. 

Finally this issue exists in at least berks 4.3.2 but *not* in *3.1.5* and actually caused us to pin to berkshelf 3.x until I finally had time to figure this out. Hope this helps someone else!

**Comment by mreeves1** [comment]
Looking more into the [code for ruby tmpdirs](http://ruby-doc.org/stdlib-2.1.2/libdoc/tmpdir/rdoc/Dir.html#tmpdir) I see this is a pretty big edge case. My /tmp dir must not have the sticky bit set... 

```
    for dir in [ENV['TMPDIR'], ENV['TMP'], ENV['TEMP'], @@systmpdir, '/tmp', '.']
      next if !dir
      dir = File.expand_path(dir)
      if stat = File.stat(dir) and stat.directory? and stat.writable? and
          (!stat.world_writable? or stat.sticky?)
        tmp = dir
        break
      end rescue nil
    end
```

**Comment by lamont-granquist** [comment]
if this is due to your /tmp not having sticky bit set it is definitely a problem with your config and not a bug here

**Comment by jimdowling** [comment]
This is still a big gotcha. I came across this problem on 2 production machines.
